WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,777 @@
+  comrvOverlayToken_t    unToken;
+  /* fixed RAM address that this cache entry refers to */
+  void*                  pFixedEntryAddress;
+} comrvCacheEntry_t;
+
+/* comrv control block */
+typedef struct comrvCB
+{
+#ifdef D_COMRV_EVICTION_LRU
+  /* holds the cache entry index of the LRU item */
+  lru_t             ucLruIndex;
+  /* holds the cache entry index of the MRU item */
+  lru_t             ucMruIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+  /* the cache entries */
+  comrvCacheEntry_t stOverlayCache[D_COMRV_NUM_OF_CACHE_ENTRIES];
+} comrvCB_t;
+
 /**
 * local prototypes
 */
+static void  comrvUpdateHeapEntryAccess      (u08_t ucEntryIndex);
+static u08_t comrvGetEvictionCandidates      (u08_t ucRequestedEvictionSize, u08_t* pEvictCandidatesList);
+static void* comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t unToken);
 
 /**
 * external prototypes
 */
+/* main comrv entry function - all overlay functions are invoked
+   through this function (the address of comrvEntry() is set in reg t6) */
+extern void  comrvEntry              (void);
+
+/* user hook functions - user application must implement the following 4 functions */
+extern void  comrvMemcpyHook         (void* pDest, void* pSrc, u32_t uiSizeInBytes);
+extern u32_t comrvCrcCalcHook        (void* pAddress, u16_t usMemSizeInBytes);
+extern void  comrvNotificationHook   (u32_t uiNotificationNum, u32_t uiToken);
+extern void* comrvLoadOvlayGroupHook (u32_t uiGroupOffset, void* pDest, u32_t uiSizeInBytes);
 
 /**
 * global variables
 */
+/* global comrv control block */
+static comrvCB_t         stComrvCB;
+/* global comrv stack pool */
+static comrvStackFrame_t stComrvStackPool[D_COMRV_CALL_STACK_DEPTH];
+
+// TODO: ronen - put pointer to tables in comrvCB_t?
+u16_t *pOverlayOffsetTable = (u16_t*)&overlayOffsetTable;
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+comrvOverlayToken_t *pOverlayMultiGroupTokensTable = (comrvOverlayToken_t*)&overlayMultiGroupTokensTable;
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+/**
+* COM-RV initialization function
+*
+* @param  pInitParams - initialization parameters
+*
+* @return none
+*/
+void comrvInit(comrvInitArgs_t* pInitArgs)
+{
+   u08_t              ucIndex;
+   void*              pBaseAddress = pInitArgs->pCacheMemoeyAddress;
+   comrvStackFrame_t* pStackPool   = stComrvStackPool;
+
+#ifndef D_COMRV_VERIFY_ARGS
+   /* verify input parameters */
+   if ((pInitArgs == NULL) || (pInitArgs->uiCacheSizeInBytes % D_COMRV_OVL_GROUP_SIZE_MIN))
+   {
+      comrvNotificationHook(D_COMRV_INVALID_INIT_PARAMS_ERR, 0);
+   }
+#endif /* D_COMRV_VERIFY_ARGS */
+
+#ifdef D_COMRV_EVICTION_LRU
+   /* initialize all heap entries */
+   for (ucIndex = 0 ; ucIndex < D_COMRV_NUM_OF_CACHE_ENTRIES ; ucIndex++)
+   {
+      /* initially each entry points to the previous and next neighbor cells */
+      stComrvCB.stOverlayCache[ucIndex].ucPrevIndex          = ucIndex-1;
```

Comment:
setting stComrvCB.stOverlayCache[ucIndex] to a local pointer increases the code size of comrvInit() by 2 bytes